onmt/translate/translator.py

```diff
@@ -343,6 +316,55 @@ def _run_encoder(self, batch, data_type):
                                .fill_(memory_bank.size(0))
         return src, enc_states, memory_bank, src_lengths
 
+    def _decode_and_generate(self, decoder_input, memory_bank, batch, data,
+                             memory_lengths, src_map=None,
+                             step=None, batch_offset=None):
+
+        if self.copy_attn:
+            # Turn any copied words to UNKs (index 0).
+            decoder_input = decoder_input.masked_fill(
+                decoder_input.gt(len(self.fields["tgt"].vocab) - 1), 0)
+
+        # Decoder forward, takes [tgt_len, batch, nfeats] as input
+        # and [src_len, batch, hidden] as memory_bank
+        # in case of inference tgt_len = 1, batch = beam times batch_size
+        # in case of Gold Scoring tgt_len = actual length, batch = 1 batch
+        dec_out, dec_attn = self.model.decoder(
+            decoder_input,
+            memory_bank,
+            memory_lengths=memory_lengths,
+            step=step)
+
+        # Generator forward.
+        if not self.copy_attn:
+            attn = dec_attn["std"]
+            log_probs = self.model.generator(dec_out.squeeze(0))
+            # returns [(batch_size x beam_size) , vocab ] when 1 step
+            # or [ tgt_len, batch_size, vocab ] when full sentence
+        else:
+            attn = dec_attn["copy"]
+            scores = self.model.generator(dec_out.view(-1, dec_out.size(2)),
+                                          attn.view(-1, attn.size(2)),
+                                          src_map)
+            # here we have scores [tgt_lenxbatch, vocab] or [beamxbatch, vocab]
+            if batch_offset is None:
+                scores = scores.view(batch.batch_size, -1, scores.size(-1))
```

Comment:
This line does the same thing as the `else` clause but is less general. It can be removed.

Reply:
My comment was invalid.